### PR TITLE
Allow inserts into tables above partition limit

### PR DESCRIPTION
**Pull request by wenleix**
Introduced in [#5396](https://github.com/prestodb/presto/pull/5396/files#diff-13aa0bdc2b255255b85a1b0763c2cb6dL1332), `HiveMetadata` needs `BucketHandler` bucket writing and execution in Hive. Current implementation get the `BucketHandler` by first obtaining
`HivePartitionResult`. This is unnecessary and cause issues when the output table has over 100,000 partitions.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 1]
```diff
@@ -1172,21 +1173,27 @@ private static Domain buildColumnDomain(ColumnHandle column, List<HivePartition>
     @Override
     public Optional<ConnectorNewTableLayout> getInsertLayout(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        HivePartitionResult hivePartitionResult = partitionManager.getPartitions(metastore, tableHandle, Constraint.alwaysTrue());
-        if (!hivePartitionResult.getBucketHandle().isPresent()) {
+        HiveTableHandle hiveTableHandle = (HiveTableHandle) tableHandle;
```
Comment (wenleix) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Here is the related code for fetching the table in [https://github.com/prestodb/presto/blob/2f148fd21a9431a67c038eee68807d67ed7829c6/presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java#L205](HivePartitionManager. getTable) (called by  `getPartitions`):

```
        Optional<Table> target = metastore.getTable(tableName.getSchemaName(), tableName.getTableName());
        if (!target.isPresent()) {
            throw new TableNotFoundException(tableName);
        }
        Table table = target.get();

        String protectMode = table.getParameters().get(ProtectMode.PARAMETER_NAME);
        if (protectMode != null && getProtectModeFromString(protectMode).offline) {
            throw new TableOfflineException(tableName, false, null);
        }

        String prestoOffline = table.getParameters().get(PRESTO_OFFLINE);
        if (!isNullOrEmpty(prestoOffline)) {
            throw new TableOfflineException(tableName, true, prestoOffline);
        }
```

Not sure if all the checks ( `protectMode` and `prestoOffline`) are required.

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: We definitely need to do those checks. I think they should be done `HiveMetadata.getTableHandle()`, so that we never return a table handle for an inaccessible table and thus we protect all code paths (all metadata operations for a specific table should operate on a handle).

Let's move those checks to a utility in `MetastoreUtil`, then call it from `getTableHandle`. We can also call the code from `HiveWriteUtils` (the `checkWritable` method basically duplicates the checks).

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Do this refactoring as a separate commit before this insert layout change.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 2]
```diff
@@ -1172,21 +1173,27 @@ private static Domain buildColumnDomain(ColumnHandle column, List<HivePartition>
...
+        SchemaTableName tableName = hiveTableHandle.getSchemaTableName();
+        Optional<Table> table = metastore.getTable(tableName.getSchemaName(), tableName.getTableName());
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: You can write this as
```java
Table table = metastore.getTable(tableName.getSchemaName(), tableName.getTableName())
        .orElseThrow(() -> new TableNotFoundException(tableName));
```
(we should use this trick in more places, though most of this code pre-dates Java 8)

Comment (wenleix) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: Great to learn!! 👍

**Comment by wenleix** [comment]
@electrum : It's failing the two offline table tests:
```
  TestHiveClient>AbstractTestHiveClient.testGetPartitionTableOffline:1132->AbstractTestHiveClient.getTableHandle:2809 » TableOffline
  TestHiveClient>AbstractTestHiveClient.testGetTableSchemaOffline:1042->AbstractTestHiveClient.getTableHandle:2809 » TableOffline
```

The current test code expects to succeed on `getTableHandle` but fail later:
```
        try (Transaction transaction = newTransaction()) {
            ConnectorMetadata metadata = transaction.getMetadata();
            ConnectorTableHandle tableHandle = getTableHandle(metadata, tableOffline);
            try {
                metadata.getTableLayouts(newSession(), tableHandle, new Constraint<>(TupleDomain.all(), bindings -> true), Optional.empty());
                fail("expected TableOfflineException");
            }
            catch (TableOfflineException e) {
                assertEquals(e.getTableName(), tableOffline);
            }
        }
```

Is this the expected behavior? Or should I change the test to expect failure earlier ? (when getting the table handler :) )

**Comment by wenleix** [comment]
Fix the HiveClient Test. Not sure why `PRODUCT_TEST` starts to fail -- it passed before. Looks unrelated :(

**Comment by wenleix** [comment]
Pass the CI now. Looks like some issues with AWS.

**Comment by wenleix** [comment]
Ping @electrum  :) .

File: presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java [hunk 3]
```diff
@@ -210,15 +209,12 @@ private Table getTable(SemiTransactionalHiveMetastore metastore, SchemaTableName
...
-        String protectMode = table.getParameters().get(ProtectMode.PARAMETER_NAME);
-        if (protectMode != null && getProtectModeFromString(protectMode).offline) {
-            throw new TableOfflineException(tableName, false, null);
-        }
-
-        String prestoOffline = table.getParameters().get(PRESTO_OFFLINE);
-        if (!isNullOrEmpty(prestoOffline)) {
-            throw new TableOfflineException(tableName, true, prestoOffline);
-        }
+        verifyOnline(
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java]: This call is short, no need to wrap the args

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 4]
```diff
@@ -215,9 +217,18 @@ public SemiTransactionalHiveMetastore getMetastore()
     public HiveTableHandle getTableHandle(ConnectorSession session, SchemaTableName tableName)
     {
         requireNonNull(tableName, "tableName is null");
-        if (!metastore.getTable(tableName.getSchemaName(), tableName.getTableName()).isPresent()) {
+        Optional<Table> table = metastore.getTable(tableName.getSchemaName(), tableName.getTableName());
+        if (!table.isPresent()) {
             return null;
         }
+
+        verifyOnline(
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: No need to wrap here

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java [hunk 5]
```diff
@@ -1127,9 +1127,8 @@ public void testGetPartitionTableOffline()
     {
         try (Transaction transaction = newTransaction()) {
             ConnectorMetadata metadata = transaction.getMetadata();
-            ConnectorTableHandle tableHandle = getTableHandle(metadata, tableOffline);
             try {
-                metadata.getTableLayouts(newSession(), tableHandle, new Constraint<>(TupleDomain.all(), bindings -> true), Optional.empty());
+                ConnectorTableHandle tableHandle = getTableHandle(metadata, tableOffline);
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java]: No need for a variable here: just call the method

**Comment by electrum** [comment]
Grammar nit in commit description -- pluralize "get":

"Current implementation get the" -> "Current implementation gets the"

**Comment by electrum** [comment]
Updated description:
```
HiveMetadata needs BucketHandler to build the insert layout. The
current implementation gets the BucketHandler by first obtaining
HivePartitionResult. This is unnecessary and cause the query to
fail when the output table has more partitions than the limit.

**Comment by electrum** [comment]
See if you can add a test for this in `TestHiveIntegrationSmokeTest`. We can lower `hive.max-partitions-per-scan` in `HiveQueryRunner` to some value like 1000. Then in a test, create a partitioned table, insert more partitions than that, then try to insert again. Also, in the same test, try to query the table and make sure we get `HIVE_EXCEEDED_PARTITION_LIMIT` (since we have no tests for this at all today).

**Comment by electrum** [comment]
Make the title of the last commit "Allow inserts into tables above partition limit"

This explains the high level problem being solved rather than the implementation.

**Comment by wenleix** [comment]
After rebase, it's failing the following tests:

```
  TestHiveClientFileMetastore>AbstractTestHiveClient.testGetTableSchemaOffline:1053 expected [1] but found [0]
  TestHiveClientTestingMetastore>AbstractTestHiveClient.testGetTableSchemaOffline:1053 expected [1] but found [0]
```

Related Pull Request: https://github.com/prestodb/presto/pull/7539

Looks like `metadata.listTableColumns` has some issues working together with Local Metastore. Will take a look into.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 6]
```diff
@@ -1150,6 +1150,79 @@ private void testInsertPartitionedTableExistingPartition(Session session, HiveSt
     }
 
     @Test
+    public void testPartitionPerScanLimit()
+            throws Exception
+    {
+        for (TestingHiveStorageFormat storageFormat : getAllTestingHiveStorageFormat()) {
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: I am not sure if we want to test it for all storage format. Seems a bit overkill?

File: presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClientLocal.java [hunk 7]
```diff
@@ -80,4 +80,7 @@ public void testGetAllTableColumnsInSchema() {}
 
     @Override
     public void testGetTableNames() {}
+
+    @Override
+    public void testGetTableSchemaOffline() {}
```
Comment (wenleix) [inline presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClientLocal.java]: Other test cases that use `metadata.listTableColumns` are override to be empty when testing against local client. So this is also added.

**Comment by wenleix** [comment]
Ready for review  @electrum

**Comment by wenleix** [comment]
@electrum  Any thoughts about the newly-added integration test? :)

**Comment by wenleix** [comment]
ping @electrum :)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 8]
```diff
@@ -1147,6 +1147,80 @@ private void testInsertPartitionedTableExistingPartition(Session session, HiveSt
     }
 
     @Test
+    public void testPartitionPerScanLimit()
+            throws Exception
+    {
+        TestingHiveStorageFormat storageFormat = new TestingHiveStorageFormat(
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: No need to wrap here

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 9]
```diff
@@ -1147,6 +1147,80 @@ private void testInsertPartitionedTableExistingPartition(Session session, HiveSt
...
+
+        TableMetadata tableMetadata = getTableMetadata(catalog, TPCH_SCHEMA, tableName);
+        assertEquals(tableMetadata.getMetadata().getProperties().get(STORAGE_FORMAT_PROPERTY), storageFormat);
+        assertEquals(tableMetadata.getMetadata().getProperties().get(PARTITIONED_BY_PROPERTY), ImmutableList.of("part"));
+
+        // insert 1200 partitions
+        for (int i = 0; i < 12; i++) {
+            assertUpdate(
+                    session,
+                    format(
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: This is hard to read. Can you extract some variables here?

Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]:
```java
assertUpdate(session, sql, 100);

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 10]
```diff
@@ -1147,6 +1147,80 @@ private void testInsertPartitionedTableExistingPartition(Session session, HiveSt
...
+                            "INSERT INTO " + tableName + " " +
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Do this as `"INSERT INTO %s"` so that you're not mixing string concatenation with formatting

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java [hunk 11]
```diff
@@ -1147,6 +1147,80 @@ private void testInsertPartitionedTableExistingPartition(Session session, HiveSt
...
+        // verify can query 1000 partitions
+        assertQuery(
+                session,
+                "SELECT count(foo) FROM " + tableName + " WHERE part < 1000",
+                "SELECT 1000");
+
+        // verify the rest 200 partitions are successfully inserted
+        assertQuery(
+                session,
+                "SELECT count(foo) FROM " + tableName + " WHERE part >= 1000 AND part < 1200",
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java]: Use `format` here to be consistent with the others below
